src/trait_solver/1.md

```diff
@@ -11,6 +11,21 @@
 {{#include ../../code/examples/stderr/trait_solver_1.stderr}}
 ```
 
-<!-- FIXME: explanation -->
+The trait implementation is for a higher ranked function pointer (`for<'a> fn`).
+But the where clause is different, there the `for<'a>` is parsed as part of the bound, so the bound is on a *not* higher-ranked function pointer.
+
+impl:
+```
+type: for<'a> fn(&'a u32)
+trait: Trait
+```
+
+bound:
+```
+type: fn(&'a u32)
+trait: Trait
+```
+
+Only higher-ranked function pointers implement the trait, so it fails to compile.
```

Comment:
I would probably explain it slightly differently, not sure which way is better.

I'd write something like

> Lifetime binders (also known as "for all" quantifiers, the `for<'a>`) make a distinct type, so an implementation for `for<'a> fn...` can't be used for `fn...`.
>
> The tricky part here is that `for<'a>` in the `where` clause is parsed as the part of the clause, and not the type. So `where for<'a> fn(&'a u32): Trait` is "for all lifetimes `'a` check that `fn(&'a u32)` implements `Trait`", rather than "check that `for<'a> fn(&'a u32)` implements `Trait`".
>
> Also note that you are allowed to add bounds on things which are not generics. The exact case in this example is not very useful, but you can also do things like `Vec<T>: Trait` or `String: Trait<T>` and so on, which can be quite useful.